Migrate Db values to MessagePack

MiguelPires opened the pull request:
## Description

Changes the state db to use msgpack values, except for column families with DbNil values (those can be changed later without breaking compatibility).

## Related issues

closes #3998

## Pull Request Checklist

- [ ] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [ ] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [ ] If submitting code, please run `mvn clean install -DskipTests` locally before committing

MiguelPires commented on engine/src/main/java/io/zeebe/engine/state/deployment/WorkflowPersistenceCache.java:
```diff
@@ -50,14 +49,15 @@
 
-  private final ColumnFamily<DbString, DbBuffer> digestByIdColumnFamily;
-  private final DbBuffer digest;
+  private final ColumnFamily<DbString, Digest> digestByIdColumnFamily;
```
I tried to do this but it's not very nice since they're used in different contexts it's not very easy to join them. The digest is only read and written during the workflow transformation but the version is often used in the engine so by uniting them we'd be constantly reading the digest unnecessarily. They're also written in different places, the digest is written during the transformation and the version is put in with the deployment. I'm not sure if that needs to stay like that or maybe it can be refactored

saig0 replied:
I don't see a problem that the properties are written and read in different places. I agree that reading the digest if it is not needed, feels not right. On the other side, having another column family may cost a lot of memory. But I'm not sure what is better here. Let's wait on the RocksDB investigation results of Chris.

MiguelPires commented on engine/src/main/java/io/zeebe/engine/state/instance/ElementInstance.java:
```diff
@@ -7,80 +7,95 @@
-  private long calledChildInstanceKey = -1L;
+import org.agrona.concurrent.UnsafeBuffer;
+
+public final class ElementInstance extends UnpackedObject implements DbValue {
```
I'm not sure about this. Unless we change the db to only take msgpack format instead of DbValues, this doesn't give us much right? And we don't want to force the db values to be msgpack since in the future we might want a column family with a different format for whatever reason (even if it has different backwards compat properties). But I agree we can delete some of the DbValue implementations that are not also DbKeys.

saig0 replied:
:+1:

MiguelPires commented on engine/src/main/java/io/zeebe/engine/state/message/SubscriptionValue.java:
```diff
@@ -0,0 +1,31 @@
+import io.zeebe.msgpack.property.ObjectProperty;
+import io.zeebe.protocol.impl.record.value.message.MessageStartEventSubscriptionRecord;
+
+public class SubscriptionValue extends UnpackedObject implements DbValue {
```
The ObjectProperty needs a record object so we need to know the type. Besides, this way if we want to add something to the state we just have to add a property, so I guess it's easier for the future

MiguelPires commented:
1) the previous implementation was already occasionally copying buffers in the wrap method, this implementation simply mirrored that. But I moved the copies out in some cases and removed them altogether in other cases
2) thanks, I fixed it
3) I think we can't change the keys since we often depend on entries being ordered. If we change the keys, the order would possibly change.

MiguelPires commented:
bors r+
